File: lib/src/exam/group_choice_question.dart
import 'package:flutter/foundation.dart';
import 'package:json_annotation/json_annotation.dart';

import 'optional_question.dart';
import 'subject_question.dart';
import 'question_rich_content.dart';

part 'group_choice_question.g.dart';

@JsonSerializable(explicitToJson: true)
class GroupChoiceQuestion extends SubjectQuestion {
  @override
  @JsonKey(includeToJson: true)
  QuestionType get type => QuestionType.groupChoice;
  final String? description;
  final String? image;
  final List<OptionalQuestion> options;
  @JsonKey(name: 'rich_content')
  final List<QuestionRichContent>? richContents;

  GroupChoiceQuestion({
    this.description,
    this.image,
    this.richContents,
    required this.options,
  });

  factory GroupChoiceQuestion.fromJson(Map<String, dynamic> json) =>
      _$GroupChoiceQuestionFromJson(json);

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is GroupChoiceQuestion &&
        other.type == type &&
        other.description == description &&
        other.image == image &&
        listEquals(other.richContents, richContents) &&
        listEquals(other.options, options);
  }

  @override
  int get hashCode =>
      Object.hash(type, description, image, richContents, options);

  @override
  Map<String, dynamic> toJson() => _$GroupChoiceQuestionToJson(this);
}

File: script/models/group_choice_question.py
from models.optional_question import OptionalQuestion
from models.subject_question import SubjectQuestion, QuestionType


class GroupChoiceQuestion(SubjectQuestion):
    def __init__(self, description: str | None, image: str | None, options: list[OptionalQuestion]):
        self.description = description
        self.image = image
        self.options = options

    @property
    def question_type(self) -> QuestionType:
        return QuestionType.GROUP_CHOICE

    def to_dict(self) -> dict[str, any]:
        return {
            "type": self.question_type.value,
            "description": self.description,
            "image": self.image,
            "options": [option.to_dict() for option in self.options]
        }
